src/cargo/ops/common_for_install_and_uninstall.rs

```diff
@@ -1,45 +1,481 @@
+                    Some((name.clone(), p))
+                } else {
+                    let p = self.v1.package_for_bin(name);
+                    Some((name.clone(), p))
+                }
+            })
+            .collect()
+    }
+
+    /// Mark that a package was installed.
+    pub fn mark_installed(
+        &mut self,
+        package: &Package,
+        bins: &BTreeSet<String>,
+        version_req: Option<String>,
+        opts: &CompileOptions<'_>,
+        target: String,
+        rustc: String,
+    ) {
+        if self.unstable_upgrade {
+            self.v2
+                .mark_installed(package, bins, version_req, opts, target, rustc)
+        }
+        self.v1.mark_installed(package, bins);
+    }
+
+    /// Save tracking information to disk.
+    pub fn save(&self) -> CargoResult<()> {
+        self.v1.save(&self.v1_lock).chain_err(|| {
+            format_err!(
+                "failed to write crate metadata at `{}`",
+                self.v1_lock.path().to_string_lossy()
+            )
+        })?;
+
+        if self.unstable_upgrade {
+            self.v2.save(self.v2_lock.as_ref().unwrap()).chain_err(|| {
+                format_err!(
+                    "failed to write crate metadata at `{}`",
+                    self.v2_lock.as_ref().unwrap().path().to_string_lossy()
+                )
+            })?;
+        }
+        Ok(())
+    }
+
+    /// Iterator of all installed binaries.
+    /// Items are `(pkg_id, bins)` where `bins` is the set of binaries that
+    /// package installed.
+    pub fn all_installed_bins(&self) -> impl Iterator<Item = (&PackageId, &BTreeSet<String>)> {
+        self.v1.v1.iter()
+    }
+
+    /// Set of binaries installed by a particular package.
+    /// Returns None if the package is not installed.
+    pub fn installed_bins(&self, pkg_id: PackageId) -> Option<&BTreeSet<String>> {
+        self.v1.v1.get(&pkg_id)
+    }
+
+    /// Remove a package from the tracker.
+    pub fn remove(&mut self, pkg_id: PackageId, bins: &BTreeSet<String>) {
+        self.v1.remove(pkg_id, bins);
+        if self.unstable_upgrade {
+            self.v2.remove(pkg_id, bins);
+        }
+    }
+}
+
 impl CrateListingV1 {
-    pub fn v1(&self) -> &BTreeMap<PackageId, BTreeSet<String>> {
-        &self.v1
+    fn package_for_bin(&self, bin_name: &str) -> Option<PackageId> {
+        self.v1
+            .iter()
+            .find(|(_, bins)| bins.contains(bin_name))
+            .map(|(pkg_id, _)| *pkg_id)
     }
 
-    pub fn v1_mut(&mut self) -> &mut BTreeMap<PackageId, BTreeSet<String>> {
-        &mut self.v1
+    fn mark_installed(&mut self, pkg: &Package, bins: &BTreeSet<String>) {
+        // Remove bins from any other packages.
+        for other_bins in self.v1.values_mut() {
+            for bin in bins {
+                other_bins.remove(bin);
+            }
+        }
+        // Remove empty metadata lines. If only BTreeMap had `retain`.
```

Comment:
The terminology of "metadata lines" (both here and below) I wasn't quite following, could that be expanded?